rocAL/source/readers/image/numpy_data_reader.cpp

```diff
@@ -0,0 +1,672 @@
+        ptr++;
+}
+
+template <size_t N>
+void NumpyDataReader::skip_char(const char*& ptr, const char (&what)[N]) {
+    if (strncmp(ptr, what, N - 1)) {
+        ERR("Found wrong symbol during parsing, expected symbol: " + std::string(what));
+        _header_parsing_failed = true;
+        return;
+    }
+    ptr += N - 1;
+}
+
+template <size_t N>
+bool NumpyDataReader::try_skip_char(const char*& ptr, const char (&what)[N]) {
+    if (!strncmp(ptr, what, N - 1)) {
+        ptr += N - 1;
+        return true;
+    } else {
+        return false;
+    }
+}
+
+template <size_t N>
+void NumpyDataReader::skip_field(const char*& ptr, const char (&name)[N]) {
+    skip_spaces(ptr);
+    CALL_AND_CHECK_FLAG(skip_char(ptr, "'"));
+    CALL_AND_CHECK_FLAG(skip_char(ptr, name));
+    CALL_AND_CHECK_FLAG(skip_char(ptr, "'"));
+    skip_spaces(ptr);
+    CALL_AND_CHECK_FLAG(skip_char(ptr, ":"));
+    skip_spaces(ptr);
+
+    error:
+        return;
+}
+
+template <typename T>
+T NumpyDataReader::parse_int(const char*& ptr) {
+    char* out_ptr = const_cast<char*>(ptr);  // strtol takes a non-const pointer
+    T value = static_cast<T>(strtol(ptr, &out_ptr, 10));
+    if (out_ptr == ptr) {
+        ERR("Parse error: expected a number.");
+        _header_parsing_failed = true;
+        return value;
+    }
+    ptr = out_ptr;
+    return value;
+}
+
+std::string NumpyDataReader::parse_string(const char*& input, char delim_start, char delim_end) {
+    if (*input++ != delim_start) {
+        ERR("Expected \'" + std::to_string(delim_start) + "\'");
+        _header_parsing_failed = true;
+        return "";
+    }
+    std::string out;
+    for (; *input != '\0'; input++) {
+        if (*input == '\\') {
+            switch (*++input) {
+                case '\\':
+                    out += '\\';
+                    break;
+                case '\'':
+                    out += '\'';
+                    break;
+                case '\t':
+                    out += '\t';
+                    break;
+                case '\n':
+                    out += '\n';
+                    break;
+                case '\"':
+                    out += '\"';
+                    break;
+                default:
+                    out += '\\';
+                    out += *input;
+                    break;
+            }
+        } else if (*input == delim_end) {
+            break;
+        } else {
+            out += *input;
+        }
+    }
+    if (*input++ != delim_end) {
+        ERR("Expected \'" + std::to_string(delim_end) + "\'");
+        _header_parsing_failed = true;
+        return "";
+    }
+    return out;
+}
+
+void NumpyDataReader::parse_header_data(NumpyHeaderData& target, const std::string& header) {
+    const char* hdr = header.c_str();
+    bool little_endian;
+    std::string typestr;
+    skip_spaces(hdr);
+    CALL_AND_CHECK_FLAG(skip_char(hdr, "{"));
+    CALL_AND_CHECK_FLAG(skip_field(hdr, "descr"));
+    typestr = parse_string(hdr);
```

Comment:
You can declare here itself right, why are we declaring in line 258, same for variable little endian

Reply:
That is becoz of the usage of the macro. In case it exits, the compiler was throwing errors that both typestr and little endian is undeclared. That's why I declared it and then set it only when we get to that point without any error
https://stackoverflow.com/questions/14274225/statement-goto-can-not-cross-variable-definition